Change `len` to uint16 for parsing parameter_desc

dmarkow opened the pull request:
This addresses #96. While Postgres's documentation on its use of signed vs unsigned integers is nearly non-existent, digging into [its source code](http://doxygen.postgresql.org/fe-misc_8c_source.html#l00272) shows that a uint16 is being used to store the number of parameters (Postgres's `getParamDescriptions` method calls into `pgGetInt` with a size of 2, which uses `uint16`).

I also [found discussions](http://www.postgresql.org/message-id/20120901160523.GD2969@momjian.us) that sometimes they use signed integers and sometimes unsigned.

Because of this uncertainty, I'm uncomfortable changing other clauses of `parse` to use unsigned integers without double checking each one.

I would LOVE to have a test for this change, but I have no clue where to even start on testing this (and it looks like there aren't any specific unit tests on the Messages module?) The only thing I can think is adding a query test with more than 32,767 parameters and ensuring it works? Let me know your thoughts and I can get something added. Thanks!

ericmj commented:
Using no test is fine. Can you look into which other messages need unsigned ints?

dmarkow commented:
Ok, I read through the libpq implementation and changed a couple more values to uint's so we should be good to merge this.

From digging into the libpq code, everything is actually transmitted from the backend as unsigned, but then some values are treated as signed and some as unsigned by the client, so it's not very straightforward to figure out what is and isn't supposed to be signed, short of coming up with test values for every message type to see what happens when you cross the overflow the signed value (which is how I found out having more than 32,767 parameters would crash).

The items I changed to unsigned all deal with value lengths/counts, so I feel confident that we wouldn't be expecting negative numbers anyway.

The one that may look odd though is `decode_row_values`. When matching against -1, we use a signed integer, but otherwise, we treat it as unsigned, which would allow us to handle a size upwards of 4GB (uint32) instead of 2GB (int32). (libpq takes this value and says "if it's greater than 0, treat it as a `size_t` (which is unsigned)). It's hard to tell for sure if PG would ever send a value over 2GB anyway, so maybe it could just be left as a signed integer, but since it's a > 0 length, I don't think there's any downside to just treating it as unsigned once we've made sure it's not -1.

ericmj commented:
Thanks for looking into this and supplying a patch @dmarkow!
